README.md
# react-native-sectioned-multi-select

[![NPM Version](https://img.shields.io/npm/v/react-native-sectioned-multi-select.svg?style=flat)](https://www.npmjs.com/package/react-native-sectioned-multi-select)

A multi (or single) select component with support for sub categories, search, chips.
It's intended for long-ish lists, as it opens in a Modal (I might make this optional in the future).

This is based on https://github.com/toystars/react-native-multiple-select.
The problems I had were that I needed it to be in a modal, because of nested ScrollViews not working on Android, and I needed to display categories with sub-categories.

## Caveats

- Currently all ids must be unique. So if you're pulling data from multiple sources, be aware that id clashes will cause (major) problems i.e both items with the same id will get selected. Looking for contributors for this - would require a bit of a refactor (see #33).
- There is no ajax support.

## Preview

![preview 1](https://github.com/renrizzolo/react-native-sectioned-multi-select/blob/master/previews/example_recording-1.gif)
![preview 2](https://github.com/renrizzolo/react-native-sectioned-multi-select/blob/master/previews/example_recording-2.gif)
![preview 3](https://github.com/renrizzolo/react-native-sectioned-multi-select/blob/master/previews/example_recording-3.gif)
![preview 4](https://github.com/renrizzolo/react-native-sectioned-multi-select/blob/master/previews/example_recording-4.gif)

## Usage

You can install this package with the following command:
`npm i react-native-sectioned-multi-select` or `yarn add react-native-sectioned-multi-select`

### Icons

The library uses icon names from `react-native-vector-icons/MaterialIcons`, however, it no longer imports the `react-native-vector-icons` library.
You should install [react-native-vector-icons](https://github.com/oblador/react-native-vector-icons) yourself, then pass the icon to the `IconRenderer` prop.

```
import Icon from 'react-native-vector-icons/MaterialIcons`
...
<SectionedMultiSelect IconRenderer={Icon} />
```

Similarly, for Expo environments you can use:

```
import {MaterialIcons} from '@expo/vector-icons';
...
<SectionedMultiSelect IconRenderer={MaterialIcons} />
```

If you prefer to use your own icon images or icon font, the `IconRenderer` prop can be used to replace the icons used.
See an example here: https://github.com/renrizzolo/react-native-sectioned-multi-select/blob/9c5f71852aef7a7ac03e7761d5dd810cd2ccef5d/exampleapp/App.js#L322-L397 (note the switch, you can refer to this to know what to map icon names to e.g if you're just passing in a different RN Vector Icons font).

### icons prop

You can now pass your own `icons` object to map the properties to a different icon library or your own `IconRenderer` component.
For each of these keys, the `IconRenderer` will be called with the corresponding `name` and `size` props.
All properties are spread to the `IconRenderer`, so you can override style or other props your icon component might use.

Here is the default icons object, which uses Material Icons names.

```
icons: {
  search: {
    name: 'search', // search input
    size: 24
  },
  arrowUp: {
    name: 'keyboard-arrow-up', // dropdown toggle
    size: 22
  },
  arrowDown: {
    name: 'keyboard-arrow-down', // dropdown toggle
    size: 22
  },
  selectArrowDown: {
    name: 'keyboard-arrow-down', // select
    size: 24
  },
  close: {
    name: 'close', // chip close
    size: 16
  },
  check: {
    name: 'check', // selected item
    size: 16
  },
  cancel: {
    name: 'cancel', // cancel button
    size: 18
  }
}
```

### Required props:

`items` | array  
`uniqueKey` | string  
`onSelectedItemsChange` | function  
`IconRenderer` | function or object

### Basic Example

```javascript
import React, { Component } from 'react';
import { View } from 'react-native';
import Icon from 'react-native-vector-icons/MaterialIcons'
import SectionedMultiSelect from 'react-native-sectioned-multi-select';

const items = [
  // this is the parent or 'item'
  {
    name: 'Fruits',
    id: 0,
    // these are the children or 'sub items'
    children: [
      {
        name: 'Apple',
        id: 10,
      },
      {
        name: 'Strawberry',
        id: 17,
      },
      {
        name: 'Pineapple',
        id: 13,
      },
      {
        name: 'Banana',
        id: 14,
      },
      {
        name: 'Watermelon',
        id: 15,
      },
      {
        name: 'Kiwi fruit',
        id: 16,
      },
    ],
  },
  {
    // next parent item
   ...
  },

];

export default class App extends Component {
  constructor() {
    super();
    this.state = {
      selectedItems: [],
    };
  }
  onSelectedItemsChange = (selectedItems) => {
    this.setState({ selectedItems });
  };

  render() {
    return (
      <View>
        <SectionedMultiSelect
          items={items}
          IconRenderer={Icon}
          uniqueKey="id"
          subKey="children"
          selectText="Choose some things..."
          showDropDowns={true}
          onSelectedItemsChange={this.onSelectedItemsChange}
          selectedItems={this.state.selectedItems}
        />
      </View>
    );
  }
}
```

### Refs / internal methods

You can do some programatic operations by setting up a ref to the component and accessing its internal methods:

```JS
// function component
import React, { useRef } from 'react'

const App = () => {
const ref = useRef(null)
  return (
    <SectionedMultiSelect
      ...
      ref={ref}
    />
  )
}

// class component
<SectionedMultiSelect
    ...
    ref={SectionedMultiSelect => this.SectionedMultiSelect = SectionedMultiSelect}
/>
```

Remove all items:

```JS
// function component
onPress={() => ref && ref.current && ref.current._removeAllItems()}
// class component
onPress={() => this.SectionedMultiSelect._removeAllItems()}
```

Toggle the modal:

```JS
// function component
onPress={() => ref && ref.current && ref.current._toggleSelector()}
// class component
onPress={() => this.SectionedMultiSelect._toggleSelector()}
```

## Recipes

See [Recipes](https://github.com/renrizzolo/react-native-sectioned-multi-select/blob/master/Recipes.md) for more detailed usage examples / customizations.

## Items

Your items should have a `uniqueKey`(default: 'id') and a `displayKey` (default: 'name').
Any child items of that item should be in `subKey`, and they will have `uniqueKey` and `displayKey` properties. As you can see from the example above, my items all have a unique `id` property and the child items is an array within the `subKey` property.

Sub categories are optional, there's no need to have `subKey` items if you don't want to.

You can also pass in extra options to individual items:

- `disabled: true` - the item will be disabled
- Icons: the component prop `iconKey` is the name of the property that individual item icons will be derived from. E.g `icon: {uri: ...}`. See full example [here](https://github.com/renrizzolo/react-native-sectioned-multi-select/blob/master/Recipes.md#item-icons)

## Props

Props, there are lots.

### Main

| Prop                        | Default | type     | Desc                                                                                                               |
| --------------------------- | ------- | -------- | ------------------------------------------------------------------------------------------------------------------ |
| items                       |         | array    | the items                                                                                                          |
| uniqueKey                   | 'id'    | string   | the unique key for your items                                                                                      |
| subKey                      | 'sub'   | string   | the array of sub items within items                                                                                |
| displayKey                  | 'name'  | string   | the key for the display name / title of the item                                                                   |
| iconKey                     |         | string   | the key for the display icon / bitmap of the item                                                                  |
| selectedItems               | []      | array    | the selected items                                                                                                 |
| onSelectedItemsChange       |         | function | function that runs when an item is toggled                                                                         |
| onSelectedItemObjectsChange |         | function | function that returns the selected items as their original objects instead of an array of ids                      |
| onCancel                    |         | function | function that runs when the cancel button is pressed                                                               |
| onConfirm                   |         | function | function that runs when the confirm button is pressed                                                              |
| onToggleSelector            |         | function | callback function that runs when the selector is toggled. receives a boolean for the open/close state of the modal |

### Options

| Prop                       | Default                   | type   | Desc                                                                                                                                                         |
| -------------------------- | ------------------------- | ------ | ------------------------------------------------------------------------------------------------------------------------------------------------------------ |
| loading                    | false                     | bool   | set the loading state, shows `loadingComponent` if true                                                                                                      |
| single                     | false                     | bool   | allow only one selection                                                                                                                                     |

CHANGELOG.md
# Changelog

## 0.10.0 - 2023-05-11

### Fixed

- Add Various Accessibility Properties #290
  - adds rolls and accessibility states for selected state, expanded state
- readme fixes #275, #287

## 0.9.1 - 2022-01-09

### Fixed

- remove accidental console log #259

## 0.9.0 - 2021-12-17

### Fixed

- Changed the way styles and colors are recalculated. Style/color prop changes will now correctly reflect to sub items. This isn't a breaking change per se but it will cause different results to previous versions when changing styles/colors dynamically.

## 0.8.4 - 2021-12-14

### Fixed

- More type related fixes
- IconRenderer prop should have been required

## 0.8.3 - 2021-12-14

### Fixed

- more type related fixes

### Added

- added exampleapp version using typescript with ItemType generic and ref methods

### Fixed

## 0.8.2 - 2021-12-13

### Added

- onChangeSearchText event #210

### Fixed

- Various types fixes

  - all props that return components are now `ReactNode`. #253
  - `icons` individual icon objects are now optional and allow any extra properties you wish to be added to your icon component. #253 #232
  - added types for some useful internal methods.

- Fix TypeError: undefined is not an object (evaluating 'item[subKey]') with unpopulated items. #237

## 0.8.1 - 2020-11-03

### Fixed

- React Native's UIManager.setLayoutAnimationEnabledExperimental causing tests fail. This version has fix for it. \n Details:

```sh
TypeError: _reactNative.UIManager.setLayoutAnimationEnabledExperimental is not a function
```

## 0.8.0 - 2020-10-01

### Fixed

- 0.7.8 introduced a bug with an attempted fix of the iconRenderer. This update addresses that and icons in general.

### Added

- the `icons` prop is an object that maps the names and sizes used for the icons (all properties are spread onto the `IconRenderer`). In conjunction with `IconRenderer` you can now easily switch out the icon library. See Icons section of Readme.

### Changed

- BREAKING: `IconRenderer` is now required (also note the casing change). The lib no longer depends on `react-native-vector-icons`. For the default Material Icons icon set, you must import Icon from `react-native-vector-icons/MaterialIcons` yourself, and pass that to the`IconRenderer` prop. See Icons section of Readme.

## 0.7.8 - 2020-09-25

### Fixed

- fix global icon assignment #203
- check for subkey here idk #173
- fix recipe #111

## 0.7.7 - 2020-04-20

### Fixed

- types definition fixes

## 0.7.6 - 2019-12-05

### Fixed

- fixed chip padding when using `hideChipRemove`

## 0.7.5 - 2019-12-05

### Added

- added `hideChipRemove` boolean prop to hide chips' remove (x) touchable #150

## 0.7.4 - 2019-12-05

### Added

- added type definitions file #63

## 0.7.3 - 2019-11-28

### Fixed

- remove `ScrollView` that wraps the `FlatList` #151 (the style prop remains as scrollView for now but the element is a View)

## 0.7.2 - 2019-10-14

### Fixed

- remove deprecated `componentWillReceiveProps` #147
- fix subitem `iconRenderer` usage #142

## 0.7.1 - 2019-08-01

### Added

- added `itemsFlatListProps` and `subItemsFlatListProps` to be able to pass in custom FlatList props object

## 0.7.0 - 2019-06-25

### Fixed

- fixed regression with backdrop / SafeAreaView. Styles: `modalWrapper` wraps the modal and `backdrop` is the backdrop view inside it.

### Added

- Added `iconRenderer` prop. Material Icons is no longer required to be installed fo the lib to work. You can pass `iconRenderer` your own icon component function.
- Added Recipes doc with more details customization examples.

### Changed

- Made readme example more concise
- cleand up example app style

## 0.6.7 - 2019-05-2

### Fixed

- fixed bug with backdropView + `modalWithTouchable` taking touches from select list.
- fixed a couple of cases where items didn't update / future proofed chip remove working inside modal

## 0.6.6 - 2019-05-1

### Added

- added `parentChipsRemoveChildren` boolean prop. If true, pressing a parent chip's remove button will remove all of its selected children.
- added `parentChipContainer` and `parentChipText` to styles, so you can style parent item chips differently.
- added `selectedIconOnLeft` boolean prop. If true, selected icons show to the left of the text. #104
- added `selectedSubItem` to styles

## 0.6.5 - 2019-02-19

### Added

- added `unselectedIconComponent` #93

## 0.6.4 - 2019-01-17

### Added

- added ability to show an icon before the item text #87. Use prop `iconKey`, then add the key to your item as either a string (to show a material icons `Icon` component by `name`), object (to show an image e.g {uri: 'imageUrl' }, or a number (that is returned from require('./localimage.png') )). use style key `itemIconStyle` to style the Icon / Image
- added `autoFocus` prop (bool) #86, to autofocus the search input when the modal is opened.

## 0.6.3 - 2018-11-27

### Added

- Add `modalWithTouchable` prop. Set to true to wrap the backdrop in a `TouchableWithoutFeedback`, which will close the selector (also fires `onToggleSelector(false)`) #76

## 0.6.1 - 2018-10-23

### Fixed


exampleapp/App.js
import React, { Component } from 'react'
import {
  Platform,
  StyleSheet,
  Text,
  View,
  ScrollView,
  Switch,
  TouchableWithoutFeedback,
  TouchableOpacity,
  ActivityIndicator,
  Dimensions,
  LayoutAnimation,
  Image,
  Appearance
} from 'react-native'
import SectionedMultiSelect from 'react-native-sectioned-multi-select'
import Icon from 'react-native-vector-icons/MaterialIcons'

const img = require('./z.jpg')

// Sorry for the mess

const items = [
  {
    title: 'Fruits',
    id: 0,

    children: [
      {
        title: 'Apple',
        id: 10
      },
      {
        title: 'Strawberry',
        id: 11
      },
      {
        title: 'Pineapple',
        id: 13
      },
      {
        title: 'Banana',
        id: 14
      },
      {
        title: 'Wátermelon',
        id: 15
      },
      {
        title: 'אבטיח',
        id: 17
      },
      {
        title: 'Raspberry',
        id: 18
      },
      {
        title: 'Orange',
        id: 19
      },
      {
        title: 'Mandarin',
        id: 20
      },
      {
        title: 'Papaya',
        id: 21
      },
      {
        title: 'Lychee',
        id: 22
      },
      {
        title: 'Cherry',
        id: 23
      },
      {
        title: 'Peach',
        id: 24
      },
      {
        title: 'Apricot',
        id: 25
      }
    ]
  },
  {
    title: 'Gèms',
    id: 1,
    icon: 'cake',
    children: [
      {
        title: 'Quartz',
        id: 26
      },
      {
        title: 'Zircon',
        id: 27
      },
      {
        title: 'Sapphirè',
        id: 28
      },
      {
        title: 'Topaz',
        id: 29
      }
    ]
  },
  {
    title: 'Plants',
    id: 2,
    icon: img,

    children: [
      {
        title: "Mother In Law's Tongue",
        id: 30
      },
      {
        title: 'Yucca',
        id: 31
      },
      {
        title: 'Monsteria',
        id: 32
      },
      {
        title: 'Palm',
        id: 33
      }
    ]
  },
  {
    title: 'No child',
    id: 34
  }
]
console.log(items)

// const items2 =
//   [{
//     title: 'Plants',
//     id: 2,
//     children: [
//       {
//         title: "Mother In Law's Tongue",
//         id: 30,
//       },
//       {
//         title: 'Yucca',
//         id: 31,
//       },
//       {
//         title: 'Monsteria',
//         id: 32,
//       },
//       {
//         title: 'Palm',
//         id: 33,
//       },

//     ],
//   }]
const items2 = []
for (let i = 0; i < 100; i++) {
  items2.push({
    id: i,
    title: `item ${i}`,
    children: [
      {
        id: `10${i}`,
        title: `child 10${i}`
      },
      {
        id: `11${i}`,
        title: `child 11${i}`
      },
      {
        id: `12${i}`,
        title: `child 12${i}`
      }
    ]
  })
}

const styles = StyleSheet.create({
  center: {
    flex: 1,
    justifyContent: 'center',
    alignItems: 'center',
    marginTop: 30
  },
  container: {
    paddingTop: 40,
    paddingHorizontal: 20
  },
  welcome: {
    fontSize: 20,
    textAlign: 'center',
    margin: 10,
    color: '#333'
  },
  border: {
    borderBottomWidth: 1,
    borderBottomColor: '#dadada',
    marginBottom: 20
  },
  heading: {
    fontSize: 24,
    fontWeight: 'bold',
    marginBottom: 5,
    marginTop: 20
  },
  label: {
    fontWeight: 'bold'
  },
  switch: {
    marginBottom: 20,
    flexDirection: 'row',
    alignItems: 'flex-end',
    justifyContent: 'space-between'
  }
})
const accentMap = {
  â: 'a',
  Â: 'A',
  à: 'a',
  À: 'A',
  á: 'a',
  Á: 'A',
  ã: 'a',
  Ã: 'A',
  ê: 'e',
  Ê: 'E',
  è: 'e',
  È: 'E',
  é: 'e',
  É: 'E',
  î: 'i',
  Î: 'I',
  ì: 'i',
  Ì: 'I',
  í: 'i',
  Í: 'I',
  õ: 'o',
  Õ: 'O',
  ô: 'o',
  Ô: 'O',
  ò: 'o',
  Ò: 'O',
  ó: 'o',
  Ó: 'O',
  ü: 'u',
  Ü: 'U',
  û: 'u',
  Û: 'U',
  ú: 'u',
  Ú: 'U',
  ù: 'u',
  Ù: 'U',
  ç: 'c',
  Ç: 'C'
}
const tintColor = '#174A87'

const Loading = (props) =>
  props.hasErrored ? (
    <TouchableWithoutFeedback onPress={props.fetchCategories}>
      <View style={styles.center}>
        <Text>oops... something went wrong. Tap to reload</Text>
      </View>
    </TouchableWithoutFeedback>
  ) : (
    <View style={styles.center}>
      <ActivityIndicator size="large" />
    </View>
  )

const Toggle = (props) => (
  <TouchableWithoutFeedback
    onPress={() => props.onPress(!props.val)}
    disabled={props.disabled}
  >
    <View style={styles.switch}>
      <Text style={styles.label}>{props.name}</Text>
      <Switch
        trackColor={tintColor}
        onValueChange={(v) => props.onPress(v)}
        value={props.val}
        disabled={props.disabled}
      />
    </View>
  </TouchableWithoutFeedback>
)

export default class App extends Component {
  constructor() {
    super()
    this.state = {
      items: null,
      loading: false,
      selectedItems: [],
      selectedItems2: [],
      selectedItemObjects: [],
      currentItems: [],
      showDropDowns: false,
      single: false,
      readOnlyHeadings: false,
      highlightChildren: false,
      selectChildren: false,
      hideChipRemove: false,
      hasErrored: false,
      isDarkMode: false
    }
    this.termId = 100
    this.maxItems = 5
  }

  componentDidMount() {
    this.pretendToLoad()
    const colorScheme = Appearance.getColorScheme()
    if (colorScheme === 'dark') {
      // Use dark color scheme
      this.setState({ isDarkMode: true })
    }
    // programatically opening the select
    // this.SectionedMultiSelect._toggleSelector()
  }

  // custom icon renderer passed to iconRenderer prop
  // see the switch for possible icon name
  // values
  icon = ({ name, size = 18, style }) => {
    // flatten the styles
    const flat = StyleSheet.flatten(style)
    // remove out the keys that aren't accepted on View
    const { color, fontSize, ...styles } = flat

    let iconComponent
    // the colour in the url on this site has to be a hex w/o hash
    const iconColor =
      color && color.substr(0, 1) === '#' ? `${color.substr(1)}/` : '/'

    const Search = (
      <Image
        source={{ uri: `https://png.icons8.com/search/${iconColor}ios/` }}
        style={{ width: size, height: size }}
      />
    )
    const Down = (
      <Image
        source={{ uri: `https://png.icons8.com/down/${iconColor}ios/` }}
        style={{ width: size, height: size }}
      />
    )
    const Up = (
      <Image
        source={{ uri: `https://png.icons8.com/up/${iconColor}ios/` }}
        style={{ width: size, height: size }}
      />
    )
    const Close = (
      <Image
        source={{ uri: `https://png.icons8.com/multiply/${iconColor}ios/` }}
        style={{ width: size, height: size }}
      />
    )

    const Check = (
      <Image
        source={{
          uri: `https://png.icons8.com/checkmark/${iconColor}android/`
        }}
        style={{ width: size / 1.5, height: size / 1.5 }}
      />
    )
    const Cancel = (
      <Image
        source={{ uri: `https://png.icons8.com/cancel/${iconColor}ios/` }}
        style={{ width: size, height: size }}
      />
    )

    switch (name) {
      case 'search':
        iconComponent = Search
        break
      case 'keyboard-arrow-up':
        iconComponent = Up
        break
      case 'keyboard-arrow-down':
        iconComponent = Down
        break
      case 'close':
        iconComponent = Close
        break
      case 'check':
        iconComponent = Check
        break
      case 'cancel':
        iconComponent = Cancel
        break
      default:
        iconComponent = null
        break
    }
    return <View style={styles}>{iconComponent}</View>
  }

  getProp = (object, key) => object && this.removerAcentos(object[key])

  rejectProp = (items, fn) => items.filter(fn)

  pretendToLoad = () => {
    this.setState({ loading: true })
    setTimeout(() => {
      this.setState({ loading: false, items })
    }, 4000)
  }

  // testing a custom filtering function that ignores accents
  removerAcentos = (s) => s.replace(/[\W\[\] ]/g, (a) => accentMap[a] || a)

  filterItems = (searchTerm, items, { subKey, displayKey, uniqueKey }) => {
    let filteredItems = []
    let newFilteredItems = []
    items.forEach((item) => {
      const parts = this.removerAcentos(searchTerm.trim()).split(
        /[[ \][)(\\/?\-:]+/
      )
      const regex = new RegExp(`(${parts.join('|')})`, 'i')
      if (regex.test(this.getProp(item, displayKey))) {
        filteredItems.push(item)
      }
      if (item[subKey]) {
        const newItem = Object.assign({}, item)
        newItem[subKey] = []
        item[subKey].forEach((sub) => {
          if (regex.test(this.getProp(sub, displayKey))) {
            newItem[subKey] = [...newItem[subKey], sub]
            newFilteredItems = this.rejectProp(
              filteredItems,
              (singleItem) => item[uniqueKey] !== singleItem[uniqueKey]
            )
            newFilteredItems.push(newItem)
            filteredItems = newFilteredItems
          }
        })
      }
    })
    return filteredItems
  }

  onSelectedItemsChange = (selectedItems) => {
    console.log(selectedItems, selectedItems.length)

    if (selectedItems.length >= this.maxItems) {
      if (selectedItems.length === this.maxItems) {
        this.setState({ selectedItems })
      }
      this.setState({
        maxItems: true
      })
      return
    }
    this.setState({
      maxItems: false
    })

    const filteredItems = selectedItems.filter(
      (val) => !this.state.selectedItems2.includes(val)
    )
    this.setState({ selectedItems: filteredItems })
  }

  onSelectedItemsChange2 = (selectedItems) => {
    const filteredItems = selectedItems.filter(
      (val) => !this.state.selectedItems.includes(val)
    )
    this.setState({ selectedItems2: filteredItems })
  }

  onConfirm = () => {
    this.setState({ currentItems: this.state.selectedItems })
  }
  onCancel = () => {
    this.SectionedMultiSelect._removeAllItems()

    this.setState({
      selectedItems: this.state.currentItems
    })
    console.log(this.state.selectedItems)
  }
  onSelectedItemObjectsChange = (selectedItemObjects) => {
    this.setState({ selectedItemObjects })
    console.log(selectedItemObjects)
  }

  onSwitchToggle = (k) => {
    const v = !this.state[k]
    this.setState({ [k]: v })
  }

  fetchCategories = () => {
    this.setState({ hasErrored: false })
    fetch('http://www.mocky.io/v2/5a5573a22f00005c04beea49?mocky-delay=500ms', {
      headers: 'no-cache'
    })
      .then((response) => response.json())
      .then((responseJson) => {
        this.setState({ cats: responseJson })
      })
      .catch((error) => {
        this.setState({ hasErrored: true })
        throw error.message
      })
  }
  filterDuplicates = (items) =>
    items.sort().reduce((accumulator, current) => {
      const length = accumulator.length
      if (length === 0 || accumulator[length - 1] !== current) {
        accumulator.push(current)
      }
      return accumulator
    }, [])

  noResults = (
    <View key="a" style={styles.center}>
      <Text>Sorry! No results...</Text>
    </View>
  )

  handleAddSearchTerm = () => {
    const searchTerm = this.SectionedMultiSelect._getSearchTerm()
    const id = (this.termId += 1)
    if (
      searchTerm.length &&
      !(this.state.items || []).some((item) => item.title.includes(searchTerm))
    ) {
      const newItem = { id, title: searchTerm }
      this.setState((prevState) => ({
        items: [...(prevState.items || []), newItem]
      }))
      this.onSelectedItemsChange([...this.state.selectedItems, id])
      this.SectionedMultiSelect._submitSelection()
    }
  }

  searchAdornment = (searchTerm) =>
    searchTerm.length ? (
      <TouchableOpacity
        style={{ alignItems: 'center', justifyContent: 'center' }}
        onPress={this.handleAddSearchTerm}
      >
        <View style={{}}>
          <Image
            source={{ uri: 'https://png.icons8.com/plus' }}
            style={{ width: 16, height: 16, marginHorizontal: 15 }}
          />
          {/*   <Icon size={18} style={{ marginHorizontal: 15 }} name="add" /> */}
        </View>
      </TouchableOpacity>
    ) : null

  renderSelectText = () => {
    const { selectedItemObjects } = this.state

    const selectText = selectedItemObjects.length
      ? `I like ${selectedItemObjects
          .map((item, i) => {
            let label = `${item.title}, `
            if (i === selectedItemObjects.length - 2)
              label = `${item.title} and `
            if (i === selectedItemObjects.length - 1) label = `${item.title}.`
            return label
          })
          .join('')}`
      : 'Select a fruit'
    return (
      <Text
        style={{
          color: this.state.isDarkMode ? 'white' : 'black',
          fontSize: 18
        }}
      >
        {selectText}
      </Text>
    )
  }

  SelectOrRemoveAll = () =>
    this.SectionedMultiSelect && (
      <TouchableOpacity
        style={{
          justifyContent: 'center',
          height: 44,
          borderWidth: 0,
          paddingHorizontal: 10,
          backgroundColor: 'darkgrey',
          alignItems: 'center'
        }}
        onPress={
          this.state.selectedItems.length
            ? this.SectionedMultiSelect._removeAllItems
            : this.SectionedMultiSelect._selectAllItems
        }
      >
        <Text style={{ color: 'white', fontWeight: 'bold' }}>
          {this.state.selectedItems.length ? 'Remove' : 'Select'} all
        </Text>
      </TouchableOpacity>
    )

  onToggleSelector = (toggled) => {
    console.log('selector is ', toggled ? 'open' : 'closed')
  }
  customChipsRenderer = (props) => {
    console.log('props', props)
    return (
      <View style={{ backgroundColor: 'yellow', padding: 15 }}>
        <Text>Selected:</Text>
        {props.selectedItems.map((singleSelectedItem) => {
          const item = this.SectionedMultiSelect._findItem(singleSelectedItem)

          if (!item || !item[props.displayKey]) return null

          return (
            <View
              key={item[props.uniqueKey]}
              style={{
                flex: 0,
                marginRight: 5,
                padding: 10,
                backgroundColor: 'orange'
              }}
            >
              <TouchableOpacity
                onPress={() => {
                  this.SectionedMultiSelect._removeItem(item)
                }}
              >
                <Text>{item[props.displayKey]}</Text>
              </TouchableOpacity>
            </View>
          )
        })}
      </View>
    )
  }
  render() {
    return (
      <ScrollView
        keyboardShouldPersistTaps="always"
        style={{
          backgroundColor: this.state.isDarkMode ? '#333' : '#f8f8f8'
        }}
        contentContainerStyle={styles.container}
      >
        <Text style={styles.welcome}>
          React native sectioned multi select example.
        </Text>
        <SectionedMultiSelect
          items={this.state.items}
          ref={(SectionedMultiSelect) =>
            (this.SectionedMultiSelect = SectionedMultiSelect)
          }
          uniqueKey="id"
          subKey="children"
          displayKey="title"
          iconKey="icon"
          autoFocus
          modalWithTouchable
          modalWithSafeAreaView
          // showCancelButton
          // headerComponent={this.SelectOrRemoveAll}
          // hideConfirm
          loading={this.state.loading}
          // filterItems={this.filterItems}
          // alwaysShowSelectText
          // customChipsRenderer={this.customChipsRenderer}
          chipsPosition="top"
          searchAdornment={(searchTerm) => this.searchAdornment(searchTerm)}
          renderSelectText={this.renderSelectText}
          // noResultsComponent={this.noResults}
          loadingComponent={
            <Loading
              hasErrored={this.state.hasErrored}
              fetchCategories={this.fetchCategories}
            />
          }
          IconRenderer={Icon}
          //  cancelIconComponent={<Text style={{color:'white'}}>Cancel</Text>}
          showDropDowns={this.state.showDropDowns}
          expandDropDowns={this.state.expandDropDowns}
          animateDropDowns={false}
          readOnlyHeadings={this.state.readOnlyHeadings}
          single={this.state.single}
          showRemoveAll
          hideChipRemove={this.state.hideChipRemove}
          selectChildren={this.state.selectChildren}
          highlightChildren={this.state.highlightChildren}
          //  hideSearch
          //  itemFontFamily={fonts.boldCondensed}
          onSelectedItemsChange={this.onSelectedItemsChange}
          onSelectedItemObjectsChange={this.onSelectedItemObjectsChange}
          onCancel={this.onCancel}
          onConfirm={this.onConfirm}
          confirmText={`${this.state.selectedItems.length}/${this.maxItems} - ${
            this.state.maxItems ? 'Max selected' : 'Confirm'
          }`}
          selectedItems={this.state.selectedItems}
          colors={{
            primary: '#5c3a9e',
            success: '#5c3a9e',
            chipColor: this.state.isDarkMode ? '#f7f7f7' : '#333'
          }}
          itemNumberOfLines={3}
          selectLabelNumberOfLines={3}
          styles={{
            // chipText: {
            //   maxWidth: Dimensions.get('screen').width - 90,
            // },
            // itemText: {
            //   color: this.state.selectedItems.length ? 'black' : 'lightgrey'
            // },
            // selectedItemText: {
            //   color: 'blue',
            // },
            // subItemText: {
            //   color: this.state.selectedItems.length ? 'black' : 'lightgrey'
            // },
            item: {
              paddingHorizontal: 10
            },
            subItem: {
              paddingHorizontal: 10
            },
            selectedItem: {
              backgroundColor: 'rgba(0,0,0,0.1)'
            },
            selectedSubItem: {
              backgroundColor: 'rgba(0,0,0,0.1)'
            },
            // selectedSubItemText: {
            //   color: 'blue',
            // },
            scrollView: { paddingHorizontal: 0 }
          }}
          // cancelIconComponent={<Icon size={20} name="close" style={{ color: 'white' }} />}
        />
        <View>
          <View style={styles.border}>
            <Text style={styles.heading}>Settings</Text>
          </View>

          <Toggle
            name="Single"
            onPress={() => this.onSwitchToggle('single')}
            val={this.state.single}
          />
          <Toggle
            name="Read only headings"
            onPress={() => this.onSwitchToggle('readOnlyHeadings')}
            val={this.state.readOnlyHeadings}
          />
          <Toggle
            name="Expand dropdowns"
            onPress={() => this.onSwitchToggle('expandDropDowns')}
            val={this.state.expandDropDowns}
            disabled={!this.state.showDropDowns}
          />
          <Toggle
            name="Show dropdown toggles"
            onPress={() => this.onSwitchToggle('showDropDowns')}
            val={this.state.showDropDowns}
          />
          <Toggle
            name="Auto-highlight children"
            onPress={() => this.onSwitchToggle('highlightChildren')}
            val={this.state.highlightChildren}
            disabled={this.state.selectChildren}
          />
          <Toggle
            name="Auto-select children"
            onPress={() => this.onSwitchToggle('selectChildren')}
            disabled={this.state.highlightChildren}
            val={this.state.selectChildren}
          />
          <Toggle
            name="Hide Chip Remove Buttons"
            onPress={() => this.onSwitchToggle('hideChipRemove')}
            val={this.state.hideChipRemove}
          />

          <TouchableWithoutFeedback
            onPress={() => this.SectionedMultiSelect._removeAllItems()}
          >
            <View style={styles.switch}>
              <Text style={styles.label}>Remove All</Text>
            </View>
          </TouchableWithoutFeedback>
        </View>
      </ScrollView>
    )
  }
}
